### PR TITLE
Support getLastValue for structs in core and fix bug in struct builder with tests

### PR DESCRIPTION
**Pull request by aljazfrancic**
## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# Description:
Support getLastValue for structs in core, enable converting data descriptors to struct types for structs, enable recursively building structs form packets, and fix bug in struct builder implementation. Some test are also added for getLastValue for structs.

### CODE REVIEW

**Comment by aljazfrancic** [comment]
How struct types are added to the type manager should be looked at.

File: core/opendaq/signal/include/opendaq/data_packet_impl.h [hunk 1]
```diff
@@ -292,111 +295,188 @@ bool DataPacketImpl<TInterface>::isDataEqual(const DataPacketPtr& dataPacket) co
...
+    return StructType(descriptor.getName(), fieldNames, fieldTypes);
+}
+
+template <typename TInterface>
+inline StructPtr DataPacketImpl<TInterface>::buildStructFromPacket(void*& addr,
+                                                                   const DataDescriptorPtr& descriptor,
+                                                                   const TypeManagerPtr& typeManager) const
+{
+    const StructTypePtr structType = createTypeFromDescriptor(descriptor);
+    typeManager.addType(structType);
```
Comment (dejancrn) [inline core/opendaq/signal/include/opendaq/data_packet_impl.h:425]: What happens if the type is already registered? Won't it crash?

File: core/opendaq/signal/include/opendaq/signal_impl.h [hunk 2]
```diff
@@ -262,6 +311,26 @@ ErrCode SignalBase<TInterface, Interfaces...>::setDescriptor(IDataDescriptor* de
...
                     valueSignalsOfDomainSignal.push_back(std::move(signalPtr));
             }
+
+            if (dataDescriptor.getSampleType() == SampleType::Struct)
+            {
+                auto typeManager = context.getTypeManager();
+                const auto name = dataDescriptor.getName();
+                const StructTypePtr structType = createTypeFromDescriptor(dataDescriptor);
+
+                if (!typeManager.hasType(name))
```
Comment (dejancrn) [inline core/opendaq/signal/include/opendaq/signal_impl.h:321]: I see an issue here with multithreaded access. But it more or less relates to making type manager threadsafe, which IMHO should have atomic function like `typeManager.tryAddType` and it should return either "Success", "ExistsSame", "ExistsDifferent" status codes.

**Review by dejancrn — APPROVED** [review]
Please check what happens if buildStructFromPacket is called twice or more

File: core/opendaq/signal/include/opendaq/data_packet_impl.h (hunk 1 above)
Comment (JakaMohorkoDS) [reply core/opendaq/signal/include/opendaq/data_packet_impl.h:425]: This code should be removed, as the type is to be registered on `setDescriptor` in SignalImpl.

File: core/opendaq/signal/include/opendaq/signal_impl.h (hunk 2 above)
Comment (aljazfrancic) [reply core/opendaq/signal/include/opendaq/signal_impl.h]: As this problem is out of scope for the given commit, I've added an [appropriate issue on Jira](https://blueberrydaq.atlassian.net/browse/TBBAS-1454).

File: docs/tests/test_signals.cpp [hunk 3]
```diff
@@ -99,7 +99,7 @@ TEST_F(SignalsTest, StructData)
     auto struct1 = builder.build();
     auto struct2 = builder.build();
 
-    auto dataDesc = builder.setSampleType(SampleType::Struct).setStructFields(List<IDataDescriptor>(struct1, struct2)).build();
+    auto dataDesc = builder.setSampleType(SampleType::Struct).setName("Struct").setStructFields(List<IDataDescriptor>(struct1, struct2)).build();
```
Comment (aljazfrancic) [inline docs/tests/test_signals.cpp:102]: Couldn't find "Corresponding document: Antora/modules/explanation/pages/signals.adoc" mentioned in test_signals.cpp (just above marked line)

**Review by JakaMohorkoDS — APPROVED** [review]

**Review by JakaMohorkoDS — APPROVED** [review]

File: core/opendaq/signal/include/opendaq/signal_impl.h [hunk 4]
```diff
@@ -262,6 +339,13 @@ ErrCode SignalBase<TInterface, Interfaces...>::setDescriptor(IDataDescriptor* de
                 if (signalPtr.assigned())
                     valueSignalsOfDomainSignal.push_back(std::move(signalPtr));
             }
+
+            if (dataDescriptor.getSampleType() == SampleType::Struct)
+            {
+                const StructTypePtr structType = createTypeFromDescriptor(dataDescriptor);
+                auto typeManager = this->context.getTypeManager();
+                addToTypeManagerRecursively(typeManager, dataDescriptor, structType);
+            }
```
Comment (JakaMohorkoDS) [inline core/opendaq/signal/include/opendaq/signal_impl.h:348]: This should likely be in a try/catch block.

File: core/opendaq/signal/include/opendaq/signal_impl.h [hunk 5]
```diff
@@ -238,6 +242,79 @@ void SignalBase<TInterface, Interfaces...>::onListenedStatusChanged(bool /*liste
 {
 }
 
+template <typename TInterface, typename... Interfaces>
+inline TypePtr SignalBase<TInterface, Interfaces...>::createTypeFromDescriptor(const DataDescriptorPtr& descriptor) const
+{
+    const auto fields = descriptor.getStructFields();
```
Comment (JakaMohorkoDS) [inline core/opendaq/signal/include/opendaq/signal_impl.h:248]: A check for `getStructFields().assigned()` would be useful here.

File: core/opendaq/signal/include/opendaq/signal_impl.h [hunk 6]
```diff
@@ -238,6 +242,79 @@ void SignalBase<TInterface, Interfaces...>::onListenedStatusChanged(bool /*liste
...
+                type = createTypeFromDescriptor(field);
+                break;
+            default:
+                type = SimpleType(CoreType::ctUndefined);
+                // TODO support string, list? + test
+        }
+        fieldNames.pushBack(field.getName());
+        fieldTypes.pushBack(type);
+    }
+    return StructType(descriptor.getName(), fieldNames, fieldTypes);
```
Comment (JakaMohorkoDS) [inline core/opendaq/signal/include/opendaq/signal_impl.h:286]: We should check for `getName().assigned()` here in all likelihood.

**Review by JakaMohorkoDS — APPROVED** [review]

**Review by JakaMohorkoDS — APPROVED** [review]

**Pull request merged**